### ISSUE TITLE
wrong energies for g-quadruplexes?

### ISSUE THREAD

**Issue by sjanssen2** [ViennaRNA/ViennaRNA#181]
Hi @RaumZeit,
I am surprised at some very stable energies for g-quadruplexes, but it looks like it is what's in your `P->gquad` table:
![image](https://user-images.githubusercontent.com/11960616/221838238-9b02084f-a560-4263-a963-6b3245b9b9b2.png)

However, the following energy is not consistent between `RNAsubopt` and `RNAeval`! Input sequence is `CGGGAGGGGCGGGUGGGAGG`, structure of concern is `.++..++...++...++...`
```
(base) lummerland x86_64 ~>uname -a
Linux lummerland 5.4.0-135-generic #152-Ubuntu SMP Wed Nov 23 20:19:22 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

(base) lummerland x86_64 ~>RNAsubopt --version
RNAsubopt 2.5.1

(base) lummerland x86_64 ~>echo "CGGGAGGGGCGGGUGGGAGG" | RNAsubopt -d0 --gquad -e 99999 -v | grep -A 10 -B 10 -F '.++..++...++...++...'
((.......)......)...   8.20
((...........)..)...   8.80
((...(...)...)..)...   9.40
((.(.....)...)..)...  10.70
(((......)...)..)...  10.20
.++....++..++..++... -27.69
.++....++.++...++... -27.69
.++...++...++..++... -27.69
.++...++..++...++... -27.69
.++..++....++..++... -27.69
.++..++...++...++... -27.69  <-- 
.+++..+++.+++.+++... -27.69
.+++.+++..+++.+++... -27.69
..++...++..++..++...   1.31
..++...++.++...++...   1.31
..++..++...++..++...   1.31
..++..++..++...++...   1.31
..++.++....++..++...   1.31
..++.++...++...++...   1.31
(.......)(......)...   6.60
(......).(......)...   6.00
```
Same input sequence and input structure yields a very different energy in `RNAeval`:
```
(base) lummerland x86_64 ~>RNAeval --version
RNAeval 2.5.1

(base) lummerland x86_64 ~>RNAeval -v -d0 --gquad

Use '&' to connect sequences that shall form a complex.
Input sequence (upper or lower case) followed by structure; @ to quit
....,....1....,....2....,....3....,....4....,....5....,....6....,....7....,....8
CGGGAGGGGCGGGUGGGAGG
.++..++...++...++...

length = 20
External loop                           :     0
Correcting for presence of structured domains
G-Quadruplex  (  2, 17) L2  ( 2, 3, 3)  :   350
CGGGAGGGGCGGGUGGGAGG
.++..++...++...++...
 energy =   3.50 kcal/mol
```
`L` of this structure is `2`, and `l` is `8`, thus the green cell should be returned. However, it is the red cell with `L=3, l=4`. Could this happen due to the presence of the structure e.g. `.+++.+++..+++.+++...`, which has `L=3, l=4`?

**Comment by RaumZeit** [comment]
Hi Stefan,
thanks for the bug report! Indeed, the exhaustive enumeration of quadruplexes in RNAsubopt does not re-compute the actual energy contributions but reports the optimal energy for the repsective subsequence instead. I've fixed that in our development branch and it will be officially available with the next release (v2.6.0). Whether further adaptations with respect to the enumeration of g-quadruplexes are necessary I have to check again.

For the odd-dangles model (-d1 or -d3), I think you might have found another bug. In your example, the first G should not contribute any dangling energy at all since it is already involved in two base pairs. I'll check the corresponding code as soon as possible (been on parental leave for the past half year). I suspect that there is some mistake in the evaluation code. The implementation is split into a g-quadruplex part and the 'regular' base pairs part, and they both do not seem to talk to each other which bases are actually available for pairing, resp. dangling.

**Comment by RaumZeit** [comment]
version 2.6.0 is on its way for release. So, I'm closing this issue now... Feel free to re-open or open another issue for additional irregularities with the g-quadruplex feature
